File: permonik-api/src/main/java/cz/incad/nkp/inprove/permonikapi/volume/Volume.java
package cz.incad.nkp.inprove.permonikapi.volume;

import cz.incad.nkp.inprove.permonikapi.audit.Auditable;
import lombok.*;
import org.apache.solr.client.solrj.beans.Field;


@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = true)
@ToString(callSuper = true)
@Getter
@Setter
public class Volume extends Auditable {

    @Field
    private String id; // UUID
    @Field
    private String barCode;
    @Field
    private String dateFrom;
    @Field
    private String dateTo;
    @Field
    private String metaTitleId; // UUID of metaTitle
    @Field
    private String subName;
    @Field
    private String mutationId; // UUID of mutation
    /*
    periodicity as string
    {
      "day": "Monday",
      "numExists": true,
      "editionId": "fd041788-b3c3-4fe9-b824-899aaad62ca3",
      "pagesCount": 0,
      "name": "Mladá fronta (TESTOVACÍ DATA)",
      "subName": "",
      "isAttachment": false
    },
    */
    @Field
    private String periodicity;
    @Field
    private Integer firstNumber;
    @Field
    private Integer lastNumber;
    @Field
    private String note;
    @Field
    private Boolean showAttachmentsAtTheEnd;
    @Field
    private String signature;
    @Field
    private String ownerId; // UUID of owner
    @Field
    private Integer year;
    @Field
    private String mutationMark;

    // Custom getter for `subName`
    public String getSubName() {
        return subName == null ? "" : subName;
    }

    // Custom getter for `note`
    public String getNote() {
        return note == null ? "" : note;
    }

    // Custom getter for `signature`
    public String getSignature() {
        return signature == null ? "" : signature;
    }

    // Custom getter for `mutationMark`
    public String getMutationMark() {
        return mutationMark == null ? "" : mutationMark;
    }

}

File: permonik-web/src/utils/volume.ts
import { TEdition } from '../schema/edition'
import { copyAuditable } from '../schema/common'
import { v4 as uuid } from 'uuid'
import { TEditableVolume, TVolume } from '../schema/volume'

export const repairVolume = (
  volume: TEditableVolume,
  editions: TEdition[]
): TVolume => {
  return {
    ...copyAuditable(volume),
    id: volume.id.length ? volume.id : uuid(),
    barCode: volume.barCode?.trim() || '',
    dateFrom: volume.dateFrom ?? '',
    dateTo: volume.dateTo ?? '',
    metaTitleId: volume.metaTitleId ?? '',
    subName: volume.subName?.trim() ?? '',
    mutationId: volume.mutationId ?? '',
    periodicity:
      volume.periodicity.map((p) => ({
        numExists: p.numExists ?? false,
        isAttachment: p.isAttachment ?? false,
        editionId:
          p.editionId ?? editions.find((pub) => pub.isDefault)?.id ?? '',
        day: p.day,
        pagesCount: Number(p.pagesCount) ?? 0,
        name: p.name?.trim() ?? '',
        subName: p.subName?.trim() ?? '',
      })) ?? [],
    firstNumber: Number(volume.firstNumber) ?? -1,
    lastNumber: Number(volume.lastNumber),
    note: volume.note?.trim() ?? '',
    showAttachmentsAtTheEnd: volume.showAttachmentsAtTheEnd ?? false,
    signature: volume.signature?.trim() ?? '',
    ownerId: volume.ownerId ?? '',
    year: Number(volume.year) ?? -1,
    mutationMark: volume.mutationMark?.trim() ?? '',
    mutationMarkNumber: volume.mutationMarkNumber?.trim() ?? '',
    mutationMarkNumberDescription:
      volume.mutationMarkNumberDescription?.trim() ?? '',
  }
}
